modules/png/src/pngreader.cpp

```diff
@@ -42,46 +42,50 @@
 
 namespace inviwo {
 
-PNGLayerReaderException::PNGLayerReaderException(const std::string& message,
-                                                 ExceptionContext context)
-    : DataReaderException(message, context) {}
-
 PNGLayerReader::PNGLayerReader() : DataReaderType<Layer>() {
     addExtension(FileExtension("png", "Portable Network Graphics"));
 }
 
 PNGLayerReader* PNGLayerReader::clone() const { return new PNGLayerReader(*this); }
 
-std::shared_ptr<inviwo::Layer> PNGLayerReader::readData(const std::string& filePath) {
-    if (!filesystem::fileExists(filePath)) throw PNGLayerReaderException(filePath);
+std::shared_ptr<inviwo::Layer> PNGLayerReader::readData(std::string_view filePath) {
+    if (!filesystem::fileExists(filePath)) throw DataReaderException(IVW_CONTEXT, filePath);
 
     auto* fp = filesystem::fopen(filePath, "rb");
-    if (!fp) throw PNGLayerReaderException("Failed to open file for reading, " + filePath);
+    if (!fp)
+        throw DataReaderException(IVW_CONTEXT, "Failed to open file for reading, {}", filePath);
     util::OnScopeExit closeFile([fp]() { fclose(fp); });
 
+    return readData(fp, filePath);
+}
+
+std::shared_ptr<inviwo::Layer> PNGLayerReader::readData(FILE* fp, std::string_view name) {
+
     unsigned char header[8];
     const auto read = fread(header, 1, 8, fp);
     if (read != 8 || png_sig_cmp(header, 0, 8)) {
-        throw PNGLayerReaderException("File is not a PNG, " + filePath);
+        throw DataReaderException(IVW_CONTEXT, "File is not a PNG, {}", name);
     }
 
     auto png_ptr = png_create_read_struct(PNG_LIBPNG_VER_STRING, NULL, NULL, NULL);
     if (!png_ptr) {
-        throw PNGLayerReaderException("Internal PNG Error: Failed to create read struct");
+        throw DataReaderException(IVW_CONTEXT,
+                                  "Internal PNG Error: Failed to create read struct: {}", name);
     }
-    util::OnScopeExit cleanup([&]() { png_destroy_read_struct(&png_ptr, NULL, NULL); });
+    util::OnScopeExit cleanup1([&]() { png_destroy_read_struct(&png_ptr, NULL, NULL); });
```

Comment:
why `cleanup1` ?

Reply:
since there is a cleanup2 below. you can not have them use the same OnScopeExit, that will delete them in thw wrong order. We have UB here before.